pkg/logql/engine.go

```diff
@@ -378,11 +378,10 @@ func (q *query) evalSample(ctx context.Context, expr syntax.SampleExpr) (promql_
 			return nil, fmt.Errorf("unsupported result type: %T", r)
 		}
 	}
-	return nil, nil
+	return promql.Vector{}, nil
```

Comment:
I wonder if we should return an empty matrix if it's a range query.

Reply:
IMO the `stepEvaluator` should return a non-nil result when it's supposed to be an empty result.

Reply:
> IMO the stepEvaluator should return a non-nil result when it's supposed to be an empty result.

That's not the case. The new step evaluator returns `false` and an empty non-nil `SampleVector`. The problem is how we handle `next` above:

```go
if next && r != nil {
```

Reply:
Or are you saying we should return `true, ts, SampleVector(vec)` [here](https://github.com/grafana/loki/blob/15c8b45d26f7dcb3181e1b5ff00796ca0bad720a/pkg/logql/first_last_over_time.go#L161-L163).